docs/src/proposals/abi-v2.md

```diff
@@ -0,0 +1,88 @@
+
+## Motivation
+The reason we need a new serialization format and thus also a new ABI is because the existing one is not extensibile and stores the account data in-place / in-stream. This prevents accounts from being memory mapped individually and also hinders reallocation of accounts (changing their size). Furthermore, there is a lot of potential for optimizations such as encoding the meta data only once per transaction and partially per instruction.
+
+## Existing Encoding Standards
+Information about each entry (attribute value pair) which is either encoded explicitly or can be implicitly gained from the context:
+- Attribute (semantics) as a tag or name string
+- Data type / encoding
+- Value length
+- Value content
+
+Some [binary data-serialization formats](https://en.wikipedia.org/wiki/Comparison_of_data-serialization_formats#Comparison_of_binary_formats) are [CBOR](https://cbor.io), [Ion](https://amzn.github.io/ion-docs/), [MsgPack](https://msgpack.org), [Protobuf](https://github.com/protocolbuffers/protobuf).
+We currently only have CBOR in our Cargo.lock, so everything else would add a dependency.
+However, none of these support indirections / references / offsets / pointers and they are all in row-major order (see next point). Therefore, we should design and implement our own extensibile serialization format.
+
+## Nesting order
+Borrowing the terms from table based database systems:
+- Row-major encoding (inner loop over attributes, outer loop over accounts) means first the values of all attributes of one account are encoded, then comes the next account.
+- Column-major encoding (inner loop over accounts, outer loop over attributes) means first the values of one attribute for all accounts are encoded, then comes the next attribute.
+
+Column-major encoding has the following advantages in our use case:
+- Extensibility on an attribute level (but not per account level as that would be wasteful).
+- Read / write access control per attribute (not per account meta data, but the account data is handled separately).
+- Splitting meta data into a local instruction context and a global transaction context.
+
+## Design
+This section describes how the new encoding works in general and how we would use it for the new ABI specifically.
+
+### Encoding
+Everything is in little endian.
+
+```Rust
+struct Map {
+  number_of_entries: u16,
+  attribute: [u16; number_of_entries],
+  value_offset: [u32; number_of_entries + 1],
+  is_value_indirect: [u8; number_of_entries],
+  values: [Value],
+}
+
+enum Value {
+  Direct([u8]),
+  Indirect({ pointer: u64, length: u32 }),
+}
+```
+
+### Memory Mapping / Regions
+Additional to the mandatory RBPF memory regions (null, program, stack, heap) we would add `2 + number_of_accounts_in_instruction` memory regions to the VM mapping.
+
+#### Read-only meta data region:
+- Transaction context:
```

Comment:
Are both the transaction context and the instruction context passed to each program?

Reply:
Yes, behind the global transaction context follows a stack with a local instruction context in each frame. All in the same memory region. That also means that deeper invocations can see their parent frames, but I don't think that is a problem, because it is public via the network protocol anyway.

Reply:
Any reason not to share the entire region (transaction context and all frames) with each layer of the call stack but pass each program an index to their instruction context within that region?

Also, what does this look like when making a CPI call?  Do we copy and modify the instruction contexts to include the current CPI instruction?

Reply:
The idea is to replace the internal CPI [`invoke_stack`](https://github.com/solana-labs/solana/blob/5bb43280eb531e4702d3ed1ddc37dd0dd65d3668/runtime/src/message_processor.rs#L290) of the `InvokeContext` by the `InvocationStack` in the encoded structure and have it operate directly on that as well.